File: src/pose_resolver.py
"""
pose_resolver — 姿态冲突裁判器

职责：给定 SpeakerReader 推荐的 mode + 当前 context，
经过 safety_hooks 抬升后，输出最终生效的 mode。

不含 LLM 调用。纯函数、确定性。
"""

from __future__ import annotations
from typing import Any


def _get_path(ctx: dict, path: str) -> Any:
    cur: Any = ctx
    for seg in path.split("."):
        if not isinstance(cur, dict):
            return None
        cur = cur.get(seg)
    return cur


def _cmp(op: str, actual: Any, expected: Any) -> bool:
    if actual is None:
        return False
    if op == "eq":
        return actual == expected
    if op == "ne":
        return actual != expected
    if op in ("gte", "lte", "gt", "lt"):
        if not (isinstance(actual, (int, float)) and isinstance(expected, (int, float))):
            return False
        if op == "gte":
            return actual >= expected
        if op == "lte":
            return actual <= expected
        if op == "gt":
            return actual > expected
        if op == "lt":
            return actual < expected
    return False


def eval_trigger(expr: dict, ctx: dict) -> bool:
    for op in ("eq", "ne", "gte", "lte", "gt", "lt"):
        if op in expr:
            return _cmp(op, _get_path(ctx, expr[op]["path"]), expr[op]["value"])
    if "any_of" in expr:
        return any(eval_trigger(e, ctx) for e in expr["any_of"])
    if "all_of" in expr:
        return all(eval_trigger(e, ctx) for e in expr["all_of"])
    if "not" in expr:
        return not eval_trigger(expr["not"], ctx)
    return False


def resolve(rules: dict, recommended: str, ctx: dict) -> dict:
    modes = rules["modes"]
    if recommended not in modes:
        raise ValueError(f"UNKNOWN_MODE: {recommended}")

    final_mode = recommended
    escalated_by = None

    for hook in rules["global"]["safety_hooks"]:
        forced = hook["forced_mode"]
        if forced not in modes:
            raise ValueError(f"UNKNOWN_MODE: {forced}")
        if eval_trigger(hook["trigger"], ctx):
            if modes[forced]["priority"] > modes[final_mode]["priority"]:
                final_mode = forced
                escalated_by = hook["name"]

    return {"mode": final_mode, "escalated_by": escalated_by}

File: src/pose_resolver.ts
/**
 * pose_resolver — 姿态冲突裁判器
 *
 * 职责：给定 SpeakerReader 推荐的 mode + 当前 context，
 * 经过 safety_hooks 抬升后，输出最终生效的 mode。
 *
 * 不含 LLM 调用。纯函数、确定性。
 */

// ── types ──

export interface TriggerCmp {
  path: string;
  value: string | number | boolean | null;
}

export type TriggerExpr =
  | { eq: TriggerCmp }
  | { ne: TriggerCmp }
  | { gte: TriggerCmp }
  | { lte: TriggerCmp }
  | { gt: TriggerCmp }
  | { lt: TriggerCmp }
  | { any_of: TriggerExpr[] }
  | { all_of: TriggerExpr[] }
  | { not: TriggerExpr };

export interface SafetyHook {
  name: string;
  trigger: TriggerExpr;
  forced_mode: string;
  src: string;
}

export interface ModeEntry {
  priority: number;
  [k: string]: unknown;
}

export interface PoseRules {
  modes: Record<string, ModeEntry>;
  global: {
    safety_hooks: SafetyHook[];
    [k: string]: unknown;
  };
  [k: string]: unknown;
}

export interface ResolverContext {
  [key: string]: unknown;
}

export interface ResolveResult {
  mode: string;
  escalated_by: string | null;
}

// ── trigger evaluator ──

function getPath(ctx: ResolverContext, path: string): unknown {
  let cur: unknown = ctx;
  for (const seg of path.split(".")) {
    if (cur == null || typeof cur !== "object") return undefined;
    cur = (cur as Record<string, unknown>)[seg];
  }
  return cur;
}

function cmp(
  op: "eq" | "ne" | "gte" | "lte" | "gt" | "lt",
  actual: unknown,
  expected: string | number | boolean | null,
): boolean {
  if (actual === undefined) return false;
  switch (op) {
    case "eq":  return actual === expected;
    case "ne":  return actual !== expected;
    case "gte": return typeof actual === "number" && typeof expected === "number" && actual >= expected;
    case "lte": return typeof actual === "number" && typeof expected === "number" && actual <= expected;
    case "gt":  return typeof actual === "number" && typeof expected === "number" && actual > expected;
    case "lt":  return typeof actual === "number" && typeof expected === "number" && actual < expected;
  }
}

function evalTrigger(expr: TriggerExpr, ctx: ResolverContext): boolean {
  if ("eq"  in expr) return cmp("eq",  getPath(ctx, expr.eq.path),  expr.eq.value);
  if ("ne"  in expr) return cmp("ne",  getPath(ctx, expr.ne.path),  expr.ne.value);
  if ("gte" in expr) return cmp("gte", getPath(ctx, expr.gte.path), expr.gte.value);
  if ("lte" in expr) return cmp("lte", getPath(ctx, expr.lte.path), expr.lte.value);
  if ("gt"  in expr) return cmp("gt",  getPath(ctx, expr.gt.path),  expr.gt.value);
  if ("lt"  in expr) return cmp("lt",  getPath(ctx, expr.lt.path),  expr.lt.value);
  if ("any_of" in expr) return expr.any_of.some((e) => evalTrigger(e, ctx));
  if ("all_of" in expr) return expr.all_of.every((e) => evalTrigger(e, ctx));
  if ("not" in expr) return !evalTrigger(expr.not, ctx);
  return false;
}

// ── resolver ──

export function resolve(
  rules: PoseRules,
  recommended: string,
  ctx: ResolverContext,
): ResolveResult {
  if (!(recommended in rules.modes)) {
    throw { code: "UNKNOWN_MODE", mode: recommended };
  }

  let finalMode = recommended;
  let escalatedBy: string | null = null;

  for (const hook of rules.global.safety_hooks) {
    if (!(hook.forced_mode in rules.modes)) {
      throw { code: "UNKNOWN_MODE", mode: hook.forced_mode };
    }
    if (evalTrigger(hook.trigger, ctx)) {
      const forced = rules.modes[hook.forced_mode];
      const current = rules.modes[finalMode];
      if (forced.priority > current.priority) {
        finalMode = hook.forced_mode;
        escalatedBy = hook.name;
      }
    }
  }

  return { mode: finalMode, escalated_by: escalatedBy };
}
